File: CppCode/M_314_BinaryTreeVerticalOrderTraversal.cpp
#include "lc_header.h"
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        map<int, vector<int>> records;
        queue<pair<int, TreeNode*>> q;
        q.emplace(0, root);
        while (!q.empty()) {
            auto p = q.front();
            if (p.second) {
                records[p.first].push_back(p.second->val);
                q.emplace(p.first - 1, p.second->left);
                q.emplace(p.first + 1, p.second->right);
            }
            q.pop();
        }

        vector<vector<int>> res;
        for (auto p: records) {
            res.emplace_back(p.second);
        }
        return res;
    }

    vector<vector<int>> verticalOrderDFS(TreeNode* root) {
        map<int, vector<int>> records;
        traverseVerticalDFS(root, 0, records);
        vector<vector<int>> res;
        for (auto p: records) {
            res.emplace_back(p.second);
        }
        return res;
    }
    void traverseVerticalDFS(TreeNode* root, int col, map<int, vector<int>>& records) {
        if (!root) return;
        records[col].push_back(root->val);
        traverseVerticalDFS(root->left, col-1, records);
        traverseVerticalDFS(root->right, col+1, records);
    }
};

File: JavaScriptCode/M_314_BinaryTreeVerticalOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalOrder = function(root) {
    let res = [];
    if (root === null) return res;
    
    let records = {};
    let queue = [];
    queue.push([0, root]);
    let min = Infinity, max = -Infinity;
    while (queue.length) {
        let p = queue.shift();
        if (p[1] !== null) {
            records[p[0]] = records[p[0]] || [];
            records[p[0]].push(p[1].val);
            min = Math.min(min, p[0]);
            max = Math.max(max, p[0]);
            queue.push([p[0]-1, p[1].left]);
            queue.push([p[0]+1, p[1].right]);
        }
    }
    while (min <= max) {
        if (records[min].length) {
            res.push(records[min]);
        }
        min++;
    }
    return res;
};